Improve scan tables in mounts

sdanielzafar opened the pull request:
## Changes
Make the `scan-tables-in-mounts` functionality much more robust and comprehensive, including adding new unit tests and fixing various bugs.

We wanted to use this for a current UC migration for a large retailer and when we ran the job originally it skipped most of the directories and found 8 total tables. With my changes we now parse many many tables. I also worked through a number of bugs.

I did my best to make minimal revisions to the existing code, but I suggest we refactor this to use `os` to list directories in the future instead of `dbutils` so that we can parallelize the operations making it scalable. 

Fixes #2540

### Functionality

- [X] modified existing workflow: scan-tables-in-mounts-experimental

### Tests

- [X] manually tested
- [X] added unit tests
- [ ] added integration tests

JCZuurmond commented on src/databricks/labs/ucx/hive_metastore/locations.py:
```diff
@@ -456,45 +456,67 @@ def _find_delta_log_folders(self, root_dir: str, delta_log_folders=None) -> dict
-                if previous_entry.is_partitioned and table_in_mount and table_in_mount.format == "DELTA":
-                    delta_log_folders[root_path] = TableInMount(format=table_in_mount.format, is_partitioned=True)
-                continue
+        if file_infos:
```
Is this necessary? Is it `None` instead of an empty list, if so, make it an empty list or return early:

```
if not file_infos:
    return delta_log_folder
```

or 

```
file_infos = self._dbutils.fs.ls(root_dir) or []
```

sdanielzafar replied:
fixed

JCZuurmond commented on src/databricks/labs/ucx/hive_metastore/locations.py:
```diff
@@ -456,45 +456,67 @@ def _find_delta_log_folders(self, root_dir: str, delta_log_folders=None) -> dict
+                    delta_log_folders[root_path] = table_in_mount
+                    logger.debug(f"Added {table_in_mount.format} table for {root_path} (general)")
+                    if self._is_recursible_dir(file_info):
+                        self._find_delta_log_folders(file_info.path, delta_log_folders)
```
Why search for more delta log folders if this one is already added?

sdanielzafar replied:
this is just part of the recursive algorithm. For instance, let's say at the user's top level directory there is a spare `test.csv` thrown in somewhere. You don't want to exit saying that the top level dir is a CSV file.

JCZuurmond commented on src/databricks/labs/ucx/hive_metastore/locations.py:
```diff
@@ -530,3 +552,23 @@ def _is_json(self, file_name: str) -> bool:
+        # - should not be the _delta_log directory
+        # - file should not be partitioned
+        # - file name should not start with part-
+        # - brackets are not allowed in dbutils.fa.ls
```
Does this make a directory non-recursible or should it be handled when calling the `ls`?

```suggestion
        # - brackets are not allowed in dbutils.fs.ls
```

sdanielzafar replied:
yes- dbutils cannot handle brackets. Something I encountered in the wild during testing on a customer's mount directory.

JCZuurmond commented on src/databricks/labs/ucx/hive_metastore/locations.py:
```diff
@@ -530,3 +552,23 @@ def _is_json(self, file_name: str) -> bool:
+            and not self._is_partitioned(file_info.name)
+            and not file_info.name.startswith("part-")
+            and not any([char in file_info.name for char in "[]"])
+            and not self._is_streaming_checkpoint_dir(file_info)
```
This is not mentioned in the comment above

JCZuurmond commented on src/databricks/labs/ucx/source_code/known.json:
```diff
@@ -237,6 +237,14 @@
     "audioread.rawread": [],
```
Remove changes in `known.json`

sdanielzafar replied:
done

JCZuurmond commented on tests/unit/hive_metastore/test_locations.py:
```diff
@@ -1,6 +1,7 @@
 from unittest.mock import Mock, call, create_autospec
 
 import pytest
+import sys
```
Move to first import block, is a built-in package

JCZuurmond commented on tests/unit/hive_metastore/test_locations.py:
```diff
@@ -306,9 +307,9 @@ def test_match_table_external_locations():
-    first_folder = FileInfo("dbfs:/mnt/test_mount/table1/", "table1/", "", "")
-    second_folder = FileInfo("dbfs:/mnt/test_mount/table2/", "table2/", "", "")
-    folder_table1 = FileInfo("dbfs:/mnt/test_mount/table1/_delta_log/", "_delta_log/", "", "")
+    first_folder = FileInfo("dbfs:/mnt/test_mount/table1/", "table1/", 0, "")
```
Should have been picked up by the linter, could you trigger the linter for this tests and the tests below by adding a type hint:

```
def test_mount_listing_multiple_folders() -> None
```

sdanielzafar replied:
I'm not understanding, test does have a type hint (return `None`)?

JCZuurmond replied:
I do not see a type hint for `test_mount_listing_multiple_folders`? 

If a method does not have a type hint in its signature then the body will not be linted

sdanielzafar replied:
@JCZuurmond - Not sure I follow. First, there are no methods here, only functions. Second, this does have a return type hint of `None` can you please be clear about what you are looking for?

https://github.com/databrickslabs/ucx/blob/9c8b22be4dd29ecba8ebb9c5ebcf9061b076cd79/tests/unit/hive_metastore/test_locations.py#L307

JCZuurmond replied:
Did you push your changes?

![Screenshot 2024-10-07 at 09 24 24](https://github.com/user-attachments/assets/9dca056a-c9f3-429c-bfcd-4bbaaac08a9f)

JCZuurmond replied:
I probably made the mistake of looking at the comment in context, thus missing that commit. Sorry for the confusion

JCZuurmond commented on tests/unit/hive_metastore/test_locations.py:
```diff
@@ -718,3 +719,184 @@ def my_side_effect(path, **_):
     assert results[2].location is None
+
+
+def test_mount_listing_misplaced_flat_file():
```
What should happen with the misplaced flat file? Should it be ignored by the tables in mounts?

sdanielzafar replied:
No, there is no simple heuristic to ignore it, so we file it as a table in case it's important.

JCZuurmond commented on tests/unit/hive_metastore/test_locations.py:
```diff
@@ -718,3 +719,184 @@ def my_side_effect(path, **_):
+    )
+
+    original_limit = sys.getrecursionlimit()
+    sys.setrecursionlimit(150)  # Temporarily lower recursion limit to catch errors early
```
Make this a fixture

nfx replied:
this may break other tests, as all of them are executed in multiple threads. not quite happy about this. is there another option?

sdanielzafar replied:
used a pytest fixure.

JCZuurmond commented on tests/unit/hive_metastore/test_locations.py:
```diff
@@ -718,3 +719,184 @@ def my_side_effect(path, **_):
+        mounts = Mounts(backend, client, "test")
+        results = TablesInMounts(backend, client, "test", mounts).snapshot()
+    except RecursionError:
+        pytest.fail("Recursion depth exceeded, possible infinite loop.")
```
Should this be an assert?

sdanielzafar replied:
no need?

JCZuurmond commented on tests/unit/hive_metastore/test_locations.py:
```diff
@@ -718,3 +719,184 @@ def my_side_effect(path, **_):
+    assert results[0].is_partitioned
+
+
+def test_mount_infinite_loop():
```
What should happen with an inifinte loop?

sdanielzafar replied:
It should be caught before going to PROD! This was a bug fix. I created this test to help identify an fix an infinite loop that sometimes appeared for files that had an `=` in them.

JCZuurmond commented on tests/unit/hive_metastore/test_locations.py:
```diff
@@ -718,3 +719,184 @@ def my_side_effect(path, **_):
+    assert len(results) == 0
+
+
+def test_mount_exclude_checkpoint_dir():
```
```suggestion
def test_mount_exclude_streaming_checkpoint():
```

OR

```suggestion
def test_mount_exclude_streaming_checkpoint_directory():
```

JCZuurmond reviewed (changes requested):
Thanks for submitting the PR. Generally looks good, have some comments.

@ericvergnaud, @nfx or @asnare  : Probably know a good way to use a UCX API for listing files on Databricks as you mention in the PR comment

sdanielzafar commented:
Thanks for the insightful comments @JCZuurmond! I have pushed some changes in between my PTO adventures. Sorry for the latency.

nfx commented on src/databricks/labs/ucx/source_code/known.json:
```diff
@@ -237,6 +237,14 @@
     "audioread.rawread": [],
     "audioread.version": []
   },
+  "autocommand": {
```
revert changes to this file

sdanielzafar replied:
done- shows this file is not being modified now, sorry abt that!

sdanielzafar commented:
@nfx - addressed your review comments, thanks! Let me know if you need anything else. I _think_ the recursion limit issue will be fixed with a pytest fixture, so I added that. If not, I'm happy to get rid of that test since it's a pointed bug fix test.

sdanielzafar commented:
@nfx - did the rebase and resolved the test issues you mentioned. I'm still getting these errors with `make fmt` which don't make sense to me as I did not modify these files. Thoughts?:
```
(.venv) dan.zafar@YFV9607WJY ucx % make fmt              
hatch run fmt
cmd [1] | black . --extend-exclude 'tests/unit/source_code/samples/'
All done! ✨ 🍰 ✨
336 files left unchanged.
cmd [2] | ruff check . --fix
All checks passed!
cmd [3] | mypy --disable-error-code 'annotation-unchecked' --exclude 'tests/unit/source_code/samples/*' --exclude dist .
venv/bin/activate_this.py:27: error: "str" has no attribute "decode"; maybe "encode"?  [attr-defined]
venv/bin/activate_this.py:30: error: Module has no attribute "real_prefix"  [attr-defined]
Found 2 errors in 1 file (checked 337 source files)
make: *** [fmt] Error 1
```

nfx reviewed (changes requested):
diff is too big and impossible to review. please rebase on the latest `main` from origin and force-push
<img width="162" alt="image" src="https://github.com/user-attachments/assets/3af6b833-2182-4f07-a69a-c118f25f7487">

sdanielzafar commented:
@nfx - was able to get the rebase sorted out after all, you can see only two files are being changed now. Can we merge this?

nfx commented on tests/unit/hive_metastore/test_locations.py:
```diff
@@ -734,3 +735,188 @@ def test_resolve_dbfs_root_in_hms_federation():
+@pytest.fixture
+def set_recursion_limit():
+    original_limit = sys.getrecursionlimit()
+    sys.setrecursionlimit(150)
```
please rewrite algorithm using the iterative approach, e.g. managing stack via list. if this hack is required on the testing side, it'll blow up in the customer side.

sdanielzafar replied:
This was just a test to catch an infinite loop in the code, which was a bug I had to solve when I picked up this code initially. I can get rid of the test if it's causing too much confusion.

JCZuurmond reviewed (commented):
LGTM besides the recursion limit, please remove and ask for a new review

nfx reviewed (approved):
let's merge this
